src/URI.cc

```diff
@@ -951,12 +989,24 @@ void URI::SetScheme(const std::string &_scheme)
 /////////////////////////////////////////////////
 URIAuthority &URI::Authority()
 {
+  if (!this->dataPtr->hasAuthority)
+  {
+    ignwarn << "Getting authority of a URI that's set not to have an authority."
+            << "Get the path instead. URI: [" << this->Str() << "]"
+            << std::endl;
+  }
```

Comment:
Since we have c++17, I think it would be safe to make this return a `std::optional<URIAuthority>` here, so that we can enforce people aren't using this incorrectly.

Reply:
My original thought was not to touch the original API, because in the long run we may want the authority to be included by default. In that case, `std::optional` may not be the right API choice in the long run. However, considering that the authority is optional even when following the correct URI spec, I agree that this can make sense in the future too.

So I removed the `HasAuthority` functions and used optional in ba2d42f